pkg/volume/csimigration/plugin_manager.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csimigration
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/component-base/featuregate"
+	csilibplugins "k8s.io/csi-translation-lib/plugins"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// PluginNameMapper contains utility methods to retrieve names of plugins
+// that support a spec, map intree <=> migrated CSI plugin names, etc
+type PluginNameMapper interface {
+	GetInTreePluginNameFromSpec(pv *v1.PersistentVolume, vol *v1.Volume) (string, error)
+	GetCSINameFromInTreeName(pluginName string) (string, error)
+}
+
+// PluginManager keeps track of migrated state of in-tree plugins
+type PluginManager struct {
+	PluginNameMapper
+}
+
+// NewPluginManager returns a new PluginManager instance
+func NewPluginManager(m PluginNameMapper) PluginManager {
+	return PluginManager{
+		PluginNameMapper: m,
+	}
+}
+
+// IsMigrationCompleteForPlugin indicates whether CSI migration has been completed
+// for a particular storage plugin
+func (pm PluginManager) IsMigrationCompleteForPlugin(pluginName string) bool {
+	// CSIMigration feature should be enabled along with the plugin-specific one
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		return false
+	}
+
+	switch pluginName {
+	case csilibplugins.AWSEBSInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAWSComplete)
+	case csilibplugins.GCEPDInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCEComplete)
+	case csilibplugins.AzureFileInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAzureFileComplete)
+	case csilibplugins.AzureDiskInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAzureDiskComplete)
+	case csilibplugins.CinderInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationOpenStackComplete)
+	default:
+		return false
+	}
+}
+
+// IsMigrationEnabledForPlugin indicates whether CSI migration has been enabled
+// for a particular storage plugin
+func (pm PluginManager) IsMigrationEnabledForPlugin(pluginName string) bool {
```

Comment:
same here, nothing in this method uses `pm`... this could be a function instead of a method, then be used by kube-controller-manager and kubelet so they don't have to maintain their own plugin -> feature name mapping

Reply:
Explained [above](https://github.com/kubernetes/kubernetes/pull/83098#discussion_r346562857)